src/raft_server.c

```diff
@@ -262,6 +268,39 @@ void raft_become_follower(raft_server_t* me_)
     me->election_timer = me->cb.get_time(me_, me->udata);
 }
 
+static int has_majority_leases(raft_server_t* me_, raft_time_t now, int grace)
+{
+    raft_server_private_t* me = (raft_server_private_t*)me_;
+    assert(me->state == RAFT_STATE_LEADER);
+
+    int n = 0;
+    int n_voting = 0;
+
+    int i;
+    for (i = 0; i < me->num_nodes; i++)
+    {
+        if (raft_node_is_voting(me->nodes[i]))
+        {
+            n_voting++;
+            if (raft_is_self(me_, me->nodes[i]) ||
+                now < raft_node_get_lease(me->nodes[i]) + grace)
+                n++;
+        }
+    }
+
+    return n_voting / 2 + 1 <= n;
+}
+
+int raft_has_majority_leases(raft_server_t* me_)
+{
+    raft_server_private_t* me = (raft_server_private_t*)me_;
+
+    if (me->state != RAFT_STATE_LEADER)
+        return 0;
+
+    return has_majority_leases(me_, me->cb.get_time(me_, me->udata), 0 /* grace */);
```

Comment:
I wrote a comment for this, but decided to delete it before pushing to GitHub. :( The `raft_has_majority_leases` function is for the user to verify if a leader has all committed entries, so that this leader may read the current state locally. For this data consistency purpose we can't be graceful. On the other hand, for the purpose of determining if a leader shall step down voluntarily because it can't maintain majority leases, we can allow a grace period, since it does not matter too much if the leader steps down a short while after the actual point when it loses majority lease.

I will make this concise and add as a comment.